File: src/lib/commands.ts
import { invoke, convertFileSrc } from "@tauri-apps/api/core";
import type { AiProvider, LogoAction, LogoResult, ProviderLogo } from "../state/types";

export interface Bounds {
  x: number;
  y: number;
  width: number;
  height: number;
}

/** 同步 AI webview：创建/显示/隐藏/销毁 */
export async function syncAiWebviews(
  providers: AiProvider[],
  activeId: string | null,
  keepState: boolean
): Promise<void> {
  await invoke("sync_ai_webviews", {
    providers: providers.map((p) => ({ id: p.id, url: p.url })),
    activeId,
    keepState,
  });
}

/** 隐藏全部 AI webview（打开设置时） */
export async function hideAiWebviews(): Promise<void> {
  await invoke("hide_ai_webviews");
}

/** 按真实内容区矩形重定位 AI webview（校正位置，不依赖硬编码侧栏宽度） */
export async function repositionAiWebviews(bounds: Bounds): Promise<void> {
  await invoke("reposition_ai_webviews", { bounds });
}

/** 各快捷键的注册结果（false = 解析失败或与系统/输入法冲突） */
export interface HotkeyRegistration {
  quickAsk: boolean;
  showMain: boolean;
  selectionToolbar: boolean;
}

/** 通知 Rust 用最新设置重新注册全局快捷键，返回每个键的注册结果 */
export async function applyHotkeys(): Promise<HotkeyRegistration> {
  return await invoke<HotkeyRegistration>("apply_hotkeys");
}

/** 显示并聚焦主窗口 */
export async function showMainWindow(): Promise<void> {
  await invoke("show_main_window");
}

/** 切换快捷提问窗显隐 */
export async function toggleQuickAsk(): Promise<void> {
  await invoke("toggle_quick_ask");
}

/** 设置快捷提问窗加载的 provider（传 url）；失败会 reject */
export async function setQuickAskProvider(url: string): Promise<void> {
  await invoke("set_quick_ask_provider", { url });
}

/** 显隐快捷提问窗的 AI 子 webview（打开 AI 选择面板时隐藏，关闭后恢复） */
export async function setQuickAskAiVisible(visible: boolean): Promise<void> {
  await invoke("set_quick_ask_ai_visible", { visible });
}

/** 隐藏快捷提问窗（顶栏隐藏按钮） */
export async function hideQuickAsk(): Promise<void> {
  await invoke("hide_quick_ask");
}

/** 设置快捷提问窗置顶（顶栏图钉按钮）；失败会 reject，供 UI 回滚 */
export async function setQuickAskPinned(pinned: boolean): Promise<void> {
  await invoke("set_quick_ask_pinned", { pinned });
}

/** 快捷提问窗新对话：导航回首页（已在首页则不操作） */
export async function quickAskNewChat(): Promise<void> {
  await invoke("quick_ask_new_chat");
}

/** LogoResult → ProviderLogo：image.path 用 convertFileSrc 转成可加载 URL */
function logoResultToProviderLogo(result: LogoResult, bustCache = false): ProviderLogo {
  if (result.type === "letter") return { type: "letter", color: result.color };
  const url = convertFileSrc(result.path);
  // 同名覆盖（替换 Logo）时文件路径不变 → URL 不变 → 浏览器命中旧图缓存、各处不刷新。
  // 上传场景追加版本号，强制 <img> 重新拉取。
  return { type: "image", src: bustCache ? `${url}?v=${Date.now()}` : url };
}

/** 新增 provider：后端生成真实 id 与 logo，前端负责写入 settings */
export async function addProvider(input: {
  name: string;
  url: string;
  enabled: boolean;
  logoAction: LogoAction;
}): Promise<{ id: string; logo: ProviderLogo }> {
  const [id, result] = await invoke<[string, LogoResult]>("add_provider", input);
  return { id, logo: logoResultToProviderLogo(result, input.logoAction.type === "upload") };
}

/** 保存已有 provider：后端校验并处理 logo，返回最终 logo 供前端写入 settings */
export async function saveProvider(input: {
  id: string;
  name: string;
  url: string;
  enabled: boolean;
  logoAction: LogoAction;
}): Promise<ProviderLogo> {
  const result = await invoke<LogoResult>("validate_and_save_provider", input);
  return logoResultToProviderLogo(result, input.logoAction.type === "upload");
}

/** 删除 provider 的 logo 文件（settings 由前端更新） */
export async function deleteProvider(id: string): Promise<void> {
  await invoke("delete_provider", { id });
}

/** 刷新主窗口当前激活的 AI webview */
export async function refreshActiveAiWebview(providerId: string): Promise<void> {
  await invoke("refresh_active_ai_webview", { providerId });
}

/** 划词捕获的待显示状态（与 Rust PendingSelection 对应） */
export interface PendingSelection {
  text: string;
  x: number;
  y: number;
  show: boolean;
}

/** 定位（防溢出）并显示划词工具条；传前端测得的逻辑尺寸 */
export async function placeAndShowSelectionToolbar(width: number, height: number): Promise<void> {
  await invoke("place_and_show_selection_toolbar", { width, height });
}

/** 隐藏划词工具条 */
export async function hideSelectionToolbar(): Promise<void> {
  await invoke("hide_selection_toolbar");
}

/** 读取划词待显示状态（首帧兜底 / 事件后读 text） */
export async function getPendingSelectionShow(): Promise<PendingSelection> {
  return await invoke<PendingSelection>("get_pending_selection_show");
}

/** 复制：把捕获文本写入剪贴板 */
export async function copySelection(): Promise<void> {
  await invoke("copy_selection");
}

/** 打开快捷提问窗口（划词「解释/翻译/总结」用，无条件显示） */
export async function showQuickAsk(): Promise<void> {
  await invoke("show_quick_ask");
}

File: src-tauri/src/commands.rs
use tauri::AppHandle;

use std::path::PathBuf;
use tauri::Manager;

use crate::provider_utils::{
    decode_png_data_url, hash_color_from_name, validate_provider_name, validate_provider_url,
    LogoAction, LogoResult,
};
use crate::quick_ask;
use crate::selection_toolbar;
use crate::shortcuts;
use crate::tray;

#[tauri::command]
pub fn apply_hotkeys(app: AppHandle) -> shortcuts::HotkeyRegistration {
    shortcuts::register_from_settings(&app)
}

#[tauri::command]
pub fn show_main_window(app: AppHandle) {
    tray::show_main(&app);
}

#[tauri::command]
pub fn toggle_quick_ask(app: AppHandle) {
    quick_ask::toggle(&app);
}

#[tauri::command]
pub fn set_quick_ask_provider(app: AppHandle, url: String) -> Result<(), String> {
    quick_ask::set_url(&app, url)
}

#[tauri::command]
pub fn set_quick_ask_ai_visible(app: AppHandle, visible: bool) -> Result<(), String> {
    quick_ask::set_ai_visible(&app, visible)
}

#[tauri::command]
pub fn hide_quick_ask(app: AppHandle) -> Result<(), String> {
    quick_ask::hide(&app)
}

#[tauri::command]
pub fn set_quick_ask_pinned(app: AppHandle, pinned: bool) -> Result<(), String> {
    quick_ask::set_pinned(&app, pinned)
}

#[tauri::command]
pub fn quick_ask_new_chat(app: AppHandle) -> Result<(), String> {
    quick_ask::new_chat(&app)
}

#[tauri::command]
pub fn place_and_show_selection_toolbar(
    app: AppHandle,
    width: f64,
    height: f64,
) -> Result<(), String> {
    selection_toolbar::place_and_show(&app, width, height)
}

#[tauri::command]
pub fn hide_selection_toolbar(app: AppHandle) -> Result<(), String> {
    selection_toolbar::hide(&app)
}

#[tauri::command]
pub fn get_pending_selection_show(app: AppHandle) -> crate::state::PendingSelection {
    selection_toolbar::get_pending(&app)
}

#[tauri::command]
pub fn copy_selection(app: AppHandle) -> Result<(), String> {
    selection_toolbar::copy_selection(&app)
}

#[tauri::command]
pub fn show_quick_ask(app: AppHandle) {
    quick_ask::show_deferred(app);
}

/// 确保 logo 目录存在并返回它
fn logo_dir(app: &AppHandle) -> Result<PathBuf, String> {
    let dir = app
        .path()
        .app_data_dir()
        .map_err(|e| e.to_string())?
        .join("provider-logos");
    std::fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    Ok(dir)
}

fn logo_path(app: &AppHandle, id: &str) -> Result<PathBuf, String> {
    Ok(logo_dir(app)?.join(format!("{id}.png")))
}

/// 处理 Logo 操作：keep 返回现有文件路径，upload 解码落盘，generate 删旧文件并返回字母色。
/// allow_keep=false 用于新增场景（不应出现 keep）。
fn apply_logo_action(
    app: &AppHandle,
    id: &str,
    action: LogoAction,
    allow_keep: bool,
) -> Result<LogoResult, String> {
    match action {
        LogoAction::Keep => {
            if !allow_keep {
                return Err("logoInvalidFormat".into());
            }
            let path = logo_path(app, id)?;
            Ok(LogoResult::Image {
                path: path.to_string_lossy().into_owned(),
            })
        }
        LogoAction::Upload { data_url } => {
            let bytes = decode_png_data_url(&data_url)?;
            let path = logo_path(app, id)?;
            std::fs::write(&path, bytes).map_err(|e| e.to_string())?;
            Ok(LogoResult::Image {
                path: path.to_string_lossy().into_owned(),
            })
        }
        LogoAction::Generate { name } => {
            // 切回字母 Logo 时清掉可能存在的旧图片文件
            let _ = std::fs::remove_file(logo_path(app, id)?);
            Ok(LogoResult::Letter {
                color: hash_color_from_name(&name),
            })
        }
    }
}

#[tauri::command]
pub fn add_provider(
    app: AppHandle,
    name: String,
    url: String,
    enabled: bool,
    logo_action: LogoAction,
) -> Result<(String, LogoResult), String> {
    validate_provider_name(&name)?;
    validate_provider_url(&url)?;
    let _ = enabled; // 启用状态由前端写入 settings，这里仅作为参数占位
    let id: String = uuid::Uuid::new_v4()
        .simple()
        .to_string()
        .chars()
        .take(8)
        .collect();
    let logo = apply_logo_action(&app, &id, logo_action, false)?;
    Ok((id, logo))
}

#[tauri::command]
pub fn validate_and_save_provider(
    app: AppHandle,
    id: String,
    name: String,
    url: String,
    enabled: bool,
    logo_action: LogoAction,
) -> Result<LogoResult, String> {
    validate_provider_name(&name)?;
    validate_provider_url(&url)?;
    // 防御性：停用时确保仍有其它启用项（前端 UI 已先拦截，此处兜底）
    if !enabled {
        let settings = crate::settings_io::read_settings(&app);
        if !crate::settings_io::other_enabled_exists(&settings.providers, &id) {
            return Err("atLeastOneEnabled".into());
        }
    }
    apply_logo_action(&app, &id, logo_action, true)
}

#[tauri::command]
pub fn delete_provider(app: AppHandle, id: String) -> Result<(), String> {
    // 防御性：删除前从 settings.json 校验删除后仍有其它启用项（前端 UI 已先禁用按钮，此处兜底）。
    // 依赖删除「前」的 settings——前端必须先调本命令、再 updateSettings，不可反序。
    let settings = crate::settings_io::read_settings(&app);
    if !crate::settings_io::other_enabled_exists(&settings.providers, &id) {
        return Err("atLeastOneEnabled".into());
    }
    // 文件可能不存在（字母 Logo），忽略 NotFound
    let _ = std::fs::remove_file(logo_path(&app, &id)?);
    Ok(())
}
